### PR TITLE
minor changes to Identity section

### PR DESCRIPTION
**Pull request by l-k-**
I tried to re-word the Identity section the way I understood it. But please ignore if this doesn't make sense -- I might have misunderstood something. I'll give it another read.

### CODE REVIEW

File: 0.1.0.md [hunk 1]
```diff
@@ -38,7 +38,9 @@ Type | Description
...
+An identity identifies a persistent thing (e.g. an object) that changes state over time. It is composed of the `domain` and `name` fields where the `domain` acts as a namespace for a set of related identities and the `name` is unique in that domain. A value may or may not have an identity. The rule of thumb is that if a value describes a volatile state of a persistent object, then this value should have an identity. Also, if the value itself needs to be "looked up" in a downstream system, then too it should have an identity.
```
Comment (bruth) [inline 0.1.0.md]: > ... then too it should have an identity.

Should it be `then it too should have an identity.`

Comment (l-k-) [inline 0.1.0.md]: fixed

**Comment by bruth** [comment]
I am going to use **entity** in place of _message_ because I think it will clarify some things.

The primary goal of the format is standardize how to model entities that change over time for transmission (e.g. streaming) or storage (queue, log, etc.). The simplest entity would only include the `value` field which may be a scalar value, a nested set of maps, a PNG (as bytes), or whatever. Adding a `name` gives the entity an identity to which the value now represents the _state_ of the entity. The `domain` field simply provides an additional level of namespacing so the namespace does not need to be included in the `name` itself, e.g. `{domain: 'dbhi', name: 'byron'}` vs. `{name: 'dbhi/byron'}`.

The `revision` field is designed to be a proxy for state of the entity since it must uniquely identity the state. Technically, if the `value` itself is not important the `revision` of the current entity (message) could be compared with the previous one for the entity to determine if it changed. If entities are explicitly versioned, the `version` field can be set to denote the version the state of the entity applies to. This is technically orthogonal to the `revision`, but related in practice.

The `refs` field provides the ability to include _soft references_ to other entities at a particular version, revision or point in time rather than copying or encoding that information in the `value`. A series of these messages with references essentially is a flattened graph. I call them _soft_ because there is no integrity checking done by the encoder, however a tool could be written to do this for a particular stream.

The `meta` field provides a place to store any other arbitrary information about the entity at that particular state. The `labels` field is simply for associating labels to entities that specify the _type_ of entity or some other classification.

**Comment by bruth** [comment]
From an tooling standpoint, we can do the following with the fields:
- Get all the domains in use
  - Index by domain
- Get all the unique entities by name
  - Index by name by domain
- Get all the unique entities by label
  - Index entities by label
- Get all the revisions of an entity
  - Index revisions by entity
- Get all version of an entity
  - Index versions by entity
- Get all the revisions for each version
  - Index entity revisions by version
- Build an entity graph
  - Create edges between entities defined in `refs`
- Build a version graph
  - Create edges between revisions and version of an entity within a domain
- Rate of values (state changes) over time
  - Uses the `time` field
- Determine if references are out of date
  - Compare a recent entity with a reference that is older than the most recent revision of the entity being reference

**Comment by l-k-** [comment]
To me, there is the entity, which is that persistent 'thing' with a unique identity, and then there is a message, or a fact, where a current/temporary value of the entity is asserted. So you would transmit/store messages which contain values of entities.
Am I reading this correctly?

**Comment by l-k-** [comment]
Will the same version values be used across all entities within a domain? That way we could get a version of the entire domain without having to follow edges defined by `refs`.  

Will refs be specified by (domain/name + version) only, or revision as well? 

Do we allow 2 entities to reference each other? What if they reference different versions?

**Comment by bruth** [comment]
Yes. The entity is identified by the identity and its state is contained in the value.

**Comment by bruth** [comment]
These points may be valuable to bake into the semantics but I need to mull over them a bit more.

**Comment by bruth** [comment]
> Will the same version values be used across all entities within a domain? That way we could get a version of the entire domain without having to follow edges defined by refs.

This is a logical constraint. A domain will likely correspond to a set/group/collection of entities that are managed together in the source system. If a producer generates facts for disparate entities that are related (cross system), this constraint would not violate that inherently since each entity would be identified under it's respective domain. Another approach a producer might take is _copying_ the entity into the domain, in which case it would inherit the domain's versioning strategy. The PEDSnet data model is actually an example of this:

``` javascript
{
    domain: "pedsnet",
    name: "person",
    labels: ["Table"],
    version: "1.0.0",
    refs: {
        derived_from: {
            domain: "omop",
            name: "person",
            version: "5.0.0"
        }
    }
}
```

> Will refs be specified by (domain/name + version) only, or revision as well?

Neither the `version` nor `revision` may be available or known for the referenced entity so we can't enforce that constraint. If a revision is provided, that is the highest level of specificity since the semantics state that it must uniquely identity the state of the entity across all time (which include all versions in the domain as well). If a `version` is specified, this is actually puts a looser constraint on which state of the referenced entity may be used. In the example above this is show and would translate to "the version 1.0.0 pedsnet person table was derived from the version 5.0.0 omop person table". This means that the latest revision of the "5.0.0 omop person table" would satisfy the reference. If no `revision` or `version` is specified, then we could either assume that any state of the referenced entity would satisfy the reference or the state of reference at the time defined by the `time` value. Alternately, the `time` field could be explicitly set on the reference to enforce this constraint.

> Do we allow 2 entities to reference each other? What if they reference different versions?

This seems valid since they could be referencing specific versions at different times. Here are few scenarios I could think of:

Mutual dependency of `A` and `B`.

``` javascript
{
    name: "A",
    revision: "1",
    time: 1,
    refs: {
        depends_on: {
            name: "B",
            revision: "1"
        }
    }
}
// ...
{
    name: "B",
    revision: "1",
    time: 1,
    refs: {
        depends_on: {
            name: "A",
            revision: "1"
        }
    }
}
```

Changes to `A` lags changes to `B`. The inverse could also be true.

``` javascript
{
    name: "A",
    revision: "1",
    time: 1,
    refs: {
        depends_on: {
            name: "B",
            revision: "1"
        }
    }
}
// ...
{
    name: "B",
    revision: "2",
    time: 2,
    refs: {
        depends_on: {
            name: "A",
            revision: "1"
        }
    }
}
```

`A` refers to a revision of `B` that occurs later (notice the `time` fields). A producer that yields this should be considered wrong since `B rev 2` doesn't yet exist, but then again that depends on the semantics of the `depends_on` relationship from `A` to `B`. This is quite a stretch, but I could imagine that `B rev 2` is known to be coming and `A rev 1` is in a _planning_ stage. That being said, I still think it leads to confusion about the relationship between the two entities.

Ultimately this is a coordination issue which is out of the scope of this format definition. If this ordering is considered to be invalid then a consumer could be added to performs validation on the stream.

``` javascript
{
    name: "A",
    revision: "1",
    time: 1,
    refs: {
        depends_on: {
            name: "B",
            revision: "2"
        }
    }
}
// ...
{
    name: "B",
    revision: "2",
    time: 2,
    refs: {
        depends_on: {
            name: "A",
            revision: "1"
        }
    }
}
```

**Comment by bruth** [comment]
Closing in favor of #2

**Pull request closed without merging**